### PR TITLE
Automate test setup using testcontainers-rs

### PR DESCRIPTION
**Pull request by stonecharioteer**
## Changes
* add dockerfile to build a nym client service
* add entrypoint script to handle variable for client ID
* add cargo build script to build the docker image every time the dockerfile changes.
* update README and tests to reflect the new way of running tests.
## Tests

Run the tests without any setup or having to compile/run nym-client ourselves.
```
cargo test
```

## Issues
Addresses #5 and builds the foundation for complete test automation.

### CODE REVIEW

File: src/transport.rs [hunk 1]
```diff
@@ -379,12 +379,34 @@ mod test {
...
+        let dialer_uri = format!("ws://0.0.0.0:{dialer_port}");
         let mut dialer_transport = NymTransport::new(&dialer_uri).await.unwrap();
-        let listener_uri = "ws://localhost:1978".to_string();
+        let nym_listener_image = GenericImage::new("nym", "latest")
+            .with_env_var("NYM_ID", "test_connection_listener")
+            .with_wait_for(nym_ready_message.clone())
+            .with_exposed_port(1977);
+        let listener_container = docker_client.run(nym_listener_image);
+        let listener_port = listener_container.get_host_port_ipv4(1977);
+        let listener_uri = format!("ws://0.0.0.0:{listener_port}");
```
Comment (noot) [inline src/transport.rs:409]: I'm surprised this works with the same client lol, how much work would it be to have docker start 2 different Nym clients on separate ports like before?

**Review by noot — APPROVED** [review]
it worked for me :D just one comment about having 2 nym clients like before, I think that'll be better for testing (and we'll need to have that setup eventually anyways)

**Comment by stonecharioteer** [comment]
Uh, that *is* two different clients and ports. Unless I'm misreading what
you were doing before?

**Comment by noot** [comment]
@stonecharioteer I only see `nym-client` being run once with port `1977`? am I missing something haha

**Comment by stonecharioteer** [comment]
Hahaha. So check the URIs. There's a dialer URI and a listener URI.

The shell script runs once for each. So when the mixnet.rs test runs, it
only launches 1 nym client and then kills it. When the transport.rs test
runs, it launches two and kills them.

You can launch as many as you need / can handle.

Check the README, I've put a code snippet there so you can write your tests
in the future. Although I'd love help to modularize it. We would
technically need to return a URI string but the container should have a
lifetime equal to that of the string. Wholesome rust fun.

**Comment by noot** [comment]
@stonecharioteer ah I get it lol, so the IPs are different for each client I guess? that makes sense! feel free to merge :D

**Comment by stonecharioteer** [comment]
Actually no, you have separate instances of nym client running within
docker. The test client library takes care of randomising port 1977, as
directed. This prevents us from hardcoding the port number everywhere.

When the tests run, type docker ps -a in another terminal to see what's
going on. The containers start and stop very quickly too, so you might want
to use the watch command.

I'll merge this. Let me know if you have any questions

**Pull request merged**